src/blockencodings.cpp

```diff
@@ -15,21 +16,45 @@
 
 #include <unordered_map>
 
+uint64_t GetShortID(
+        const uint64_t& shorttxidk0,
+        const uint64_t& shorttxidk1,
+        const uint256& txhash)
+{
+    static_assert(CompactBlock::SHORTTXIDS_LENGTH == 6, "shorttxids calculation assumes 6-byte shorttxids");
+    return SipHashUint256(shorttxidk0, shorttxidk1, txhash) & 0xffffffffffffL;
+}
+
 #define MIN_TRANSACTION_SIZE (::GetSerializeSize(CTransaction(), SER_NETWORK, PROTOCOL_VERSION))
 
-CBlockHeaderAndShortTxIDs::CBlockHeaderAndShortTxIDs(const CBlock& block) :
-        nonce(GetRand(std::numeric_limits<uint64_t>::max())),
-        shorttxids(block.vtx.size() - 1), prefilledtxn(1), header(block) {
+CompactBlock::CompactBlock(const CBlock& block,
+    const CRollingBloomFilter* inventoryKnown) :
+        nonce(GetRand(std::numeric_limits<uint64_t>::max())), header(block)
+{
     FillShortTxIDSelector();
-    //TODO: Use our mempool prior to block acceptance to predictively fill more than just the coinbase
-    prefilledtxn[0] = {0, block.vtx[0]};
+
+    if (block.vtx.empty())
+        throw std::invalid_argument(__func__ + std::string(" expects coinbase tx"));
+
+    //< Index of a prefilled tx is its diff from last index.
+    size_t prevIndex = 0;
+    prefilledtxn.push_back(PrefilledTransaction{0, block.vtx[0]});
+
     for (size_t i = 1; i < block.vtx.size(); i++) {
         const CTransaction& tx = block.vtx[i];
-        shorttxids[i - 1] = GetShortID(tx.GetHash());
+
+        if (inventoryKnown && !inventoryKnown->contains(tx.GetHash()))
```

Comment:
I failed to mention, not just blocks only peers will get spammed, but any peer you recently connected to-- probably for the first couple hours after the connection.   Which is why I think it should just check to see if it's predicted more than a few transactions or kilobytes, and if so-- send the coinbase only.